src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -108,3 +113,156 @@ def snapshot(self) -> Iterable[JobInfo]:
+
+    def snapshot(self) -> Iterable[SubmitRunInfo]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    @staticmethod
+    def dt_to_ms(date_time: datetime):
+        epoch = datetime.utcfromtimestamp(0)
+        delta = date_time - epoch
+        return int(delta.total_seconds() * 1000)
+
+    @staticmethod
+    def get_current_dttm() -> datetime:
+        return datetime.now()
+
+    def _crawl(self) -> Iterable[SubmitRunInfo]:
+        end = self.get_current_dttm()
+        start = end - timedelta(days=self._num_days_history)
+        submit_runs = self._ws.jobs.list_runs(
+            expand_tasks=True,
+            completed_only=True,
+            run_type=ListRunsRunType.SUBMIT_RUN,
+            start_time_from=self.dt_to_ms(start),
+            start_time_to=self.dt_to_ms(end),
+        )
+        return self._assess_job_runs(submit_runs)
+
+    def _try_fetch(self) -> Iterable[SubmitRunInfo]:
+        for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
+            yield SubmitRunInfo(*row)
+
+    def _check_spark_conf(self, conf: dict[str, str], source: str) -> list[str]:
+        failures: list[str] = []
+        for key in conf.keys():
+            if any(pattern in key for pattern in self._FS_LEVEL_CONF_SETTING_PATTERNS):
+                failures.append(f"Potentially unsupported config property: {key}")
+
+        failures.extend(super()._check_spark_conf(conf, source))
+        return failures
+
+    def _check_cluster_failures(self, cluster: ClusterDetails, source: str) -> list[str]:
+        failures: list[str] = []
+        if cluster.aws_attributes and cluster.aws_attributes.instance_profile_arn:
+            failures.append(f"using instance profile: {cluster.aws_attributes.instance_profile_arn}")
+
+        failures.extend(super()._check_cluster_failures(cluster, source))
+        return failures
+
+    @staticmethod
+    def _needs_compatibility_check(spec: compute.ClusterSpec) -> bool:
+        """
+        # we recognize a task as a potentially incompatible one if:
+        # 1. cluster is not configured with data security mode
+        # 2. cluster's DBR version is greater than 11.3
+        """
+        if spec.data_security_mode is None:
+            compatibility = spark_version_compatibility(spec.spark_version)
+            return compatibility == "supported"
+        return False
+
+    def _flatten_task(self, task_dict):
+        flat_dict = {}
+        for key, value in task_dict.items():
+            if isinstance(value, (str, int, float)):
+                flat_dict[key] = value
+            # convert iterable
+            if isinstance(value, collections.abc.Iterable):
+                flat_dict[key] = tuple(list(value))
+            # recursion if nested
+            elif isinstance(value, dict):
+                flat_dict[key] = self._flatten_task(value)
+        return dict(sorted(flat_dict.items(), key=lambda item: item[0]))
```

Comment:
can you bring back the hash computation function from before? this refactoring of yours does not create a deterministic hash.

Reply:
Done